MH-12759 More Control Over Workflows

krinnewitz opened the pull request:
This patch adds the possibility to stop running workflows as
well as deleting old workflows from within the admin UI.

Previously, it was already possible to resume events which were in PAUSED
state due to failed operations. Resuming could be done with or without
retrying the failed operation.

Just like these actions, the new ones can be found in the "Actions"
column of the workflow table under "Event Details" -> "Workflows":

![screenshot-2018-3-21 opencast](https://user-images.githubusercontent.com/1727976/37701976-5a043492-2cf1-11e8-80d3-e2e48def1322.png)

While deleting workflows should only be relevant to admins,
regular users might find it helpful to be able to stop workflows. Hence,
deleting workflows requires the correspong access role.

This work is sponsored by SWITCH.

FYI: The "Resume" feature was introduced by [this](https://bitbucket.org/opencast-community/opencast/pull-requests/1545/mh-12244-as-an-administrator-i-would-like/diff) PR.

Commit:
MH-12759 More Control Over Workflows

This patch adds the possibility to stop running workflows as well as
deleting old workflows from within the admin UI.

Previously, it was already possible to resume events which were in PAUSE
state due to failed operations. Resuming could be done with or without
retrying the failed operation.

Just like these actions, the new ones can be found in the "Actions"
column of the workflow table under "Event Details" -> "Workflows".

While deleting workflows should only be relevant to admins, regular
users might find it helpful to be able to stop workflows. Hence,
deleting workflows requires the correspong access role.

staubesv commented:
Hi @krinnewitz, the icons don't seem to be of the same size (e.g. delete is larger than pause and stop) and the alignment of the icons isn't correct (e.g. 2nd screenshot: the stop icon is "more right" than the delete icon on the line below).

Could you please fix this?

As for the circle icons: They should have the same size as the circle icons on Events Table->Actions

karendolan commented:
Hi  @krinnewitz, I added a comment on the MH-12759 ticket. Can you clarify the use cases for this feature. Specifically, the motivation of pausing a workflow in order to update metadata was a concern in v1x and a motivating factor to remove the PAUSED state in OCv2x. I am not against reintroducting the PAUSED state, I was not excited about removing it in 2x. But, I am concerned about reintroducting the same issues that caused it to be removed in 2x.

krinnewitz commented:
Hi @karendolan, the use case of manually pausing workflows could be an overloaded system where an admin might want to temporarily pause certain workflows. This PR ensures that event meta data cannot be edited for events in PAUSED state.

staubesv commented:
Hi @karendolan,

I've tested the following cases:

1. Workflow cannot be started on event with paused workflow using Actions->Start Task
2. Workflow cannot be started on event with paused workflow using Video Editor
3. Metadata cannot be edited on event with paused workflow using Event Details

Have I missed a case the causes problems?

Anyway: As mentioned, we could live without the Pause feature... maybe you take the review?

rute-santos commented:
I haven't yet looked at the code, but I have some comments regarding the UI:

As a user, I would like to be able to differentiate processing that was paused by the user from processing that was paused due to an error. I would prefer only seeing the warning icon (!) on the event list view when an error had occurred.

When the workflow is in paused state, the user sees 3 icons in the workflow details: stop, hand (abort), and retry. It is not clear to me what the difference between the first two (stop and abort) is.
I think it would be more clear if there were only 2 actions to choose from:
* If the processing in error: retry and abort (as before)
* If the processing has been paused by the user: stop and resume (maybe a new icon for resume?)

staubesv commented:
Hi @krinnewitz and @rute-santos: I also saw when testing that editing metadata was disabled but no message indicated that... so in overall, I think it would take quite a while to get it done when considering all the subtleties.

@krinnewitz : As "pausing workflows" is not a priority for us, please remove the pause functionality from the front-end. I think it is ok when the Admin UI Facade still can handle it, no need to remove it from there.

krinnewitz commented:
I removed the pause feature. However, events can still go into PAUSED state when workflow operations with retry strategy "hold" fail. In this case, you now see the hand-stop icon, the refresh icon and the stop icon.

I think in this case, it should still not be possible to edit meta data, which is why I kept the corresponding change to lock the meta data for PAUSED events in the commit. 

The hand icon and the stop icon might still confuse users. In my opinion, the wording and the icon for the "abort" feature is misleading. Clicking the hand-stop icon doesn't really abort a workflow, but resumes it whithout retrying failed operations (which will finally fail the complete workflow in most cases).

It is quite hard to find a suitable icon for "resume without retrying failed operations".

staubesv commented:
@krinnewitz I also think that the stop icon will not confuse users as its meaning seems "obvious". The hand icon is indeed a non-standard icon which does confuse users. That problem is, however, not part of our work, so just ignore that.

@rute-santos Kim has removed the pause function which would indeed need more work. Please continue the review.

rute-santos commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:
```diff
@@ -1014,7 +1017,8 @@ public Response getEventMetadata(@PathParam("eventId") String eventId) throws Ex
             EventUtils.getEventMetadata(optEvent.get(), getIndexService().getCommonEventCatalogUIAdapter()));
 
-    if (WorkflowInstance.WorkflowState.RUNNING.toString().equals(optEvent.get().getWorkflowState()))
+    final String wfSate = optEvent.get().getWorkflowState();
```
Typo in variable name: wfSate.

rute-santos commented:
@krinnewitz, @staubesv,

I agree that no metadata editing should be allowed when the workflow is paused.

I also agree that 'abort' is not the best name but I still find it confusing for the user to know the difference between 'stop' and 'abort' when the workflow is paused due to error.

So how about this: When the workflow is paused due to error, only have 'retry' and 'stop' as choices i.e. either re-run the operation that failed or stop the workflow?
The rationale behind this idea is that if the operation is not critical for completion of the workflow (fail-on-error="false"), then it makes more sense for it to have retry-strategy="retry" than retry-strategy="hold". What do you think?

If we agree on this, then the documentation in the admin guide (under Retry Strategies) should also be updated to reflect the change in behavior.

staubesv commented:
@karendolan For me, removing the "Abort" button seems to make sense as stop is easier to understand.

karendolan commented:
@staubesv I think you meant to direct the comment to @rute-santos! But it sounds good to me, too.

krinnewitz commented:
Hi @rute-santos, thanks for your review. I have just removed the "abort" feature. I have also removed the code you introduced in the WorkflowService to support this feature. Hopefully, I didn't screw anything up. Maybe you can pay some special attention to the WorkflowService changes while reviewing. I think you have some deeper insights there :-)

rute-santos commented:
Hi @krinnewitz, @staubesv,
@JamesUoM had a good use case for keeping an 'abort' equivalent functionality (see ticket):
"FAILING a workflow may be more useful than just STOPPED, the reason is we've implemented the archive operation in our error-handler workflow so if compose fails after editing the smil file is not lost etc."
I believe this could be easily implemented by keeping the 'NONE' strategy as before and creating the error-resolution operation with fail-on-error=true to make it fail the workflow. What do you think?
We still have the problem of finding a good icon for 'fail'...

staubesv commented:
Hi @rute-santos

> I believe this could be easily implemented by keeping the 'NONE' strategy as before and creating the error-resolution operation with fail-on-error=true to make it fail the workflow. What do you think?

Not in the scope of this ticket, but it would make sense to not remove the 'NONE' strategy then.

I suggest the following: @krinnewitz reverts the removal of the "None" Action. To solve the usability problem with users being present two similar icons for "paused" workflows, we simply don't show the stop icon in case that the workflow is paused. As for the moment, the only reason a workflow can be paused seems the retry strategy and in this case, we don't need the ability to stop a workflow as it is actually not "paused" but rather "failing".

What do you think?

rute-santos commented:
Hi @staubesv,
Sounds good to me!

Commit:
Don't show stop icon when event is 'PAUSED'

krinnewitz commented:
Ok, removal of 'abort' feature is now reverted and 'stop' button is hidden when event is in PAUSED state.

rute-santos commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:
```diff
@@ -2321,51 +2325,93 @@ public JObject apply(Opt<Recording> recording) {
+  @DELETE
+  @Path("{eventId}/workflows/{workflowId}")
+  @RestQuery(name = "deleteWorkflow", description = "Deletes a workflow", returnDescription = "The method doesn't return any content", pathParameters = {
+    @RestParameter(name = "eventId", isRequired = true, description = "The event identifier", type = RestParameter.Type.STRING),
```
Change transitionId to workflowId?

krinnewitz replied:
Oh yes, right. Done.

rute-santos commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:
```diff
@@ -2321,51 +2325,93 @@ public JObject apply(Opt<Recording> recording) {
+  @Path("{eventId}/workflows/{workflowId}")
+  @RestQuery(name = "deleteWorkflow", description = "Deletes a workflow", returnDescription = "The method doesn't return any content", pathParameters = {
+    @RestParameter(name = "eventId", isRequired = true, description = "The event identifier", type = RestParameter.Type.STRING),
+    @RestParameter(name = "transitionId", isRequired = true, description = "The workflow identifier", type = RestParameter.Type.INTEGER) }, reponses = {
```
Please add badRequest.

krinnewitz replied:
Yes, missed that one. Done.

rute-santos commented on modules/admin-ui/src/main/webapp/scripts/modules/events/controllers/eventController.js:
```diff
@@ -882,15 +882,60 @@ angular.module('adminNg.controllers')
+                        case 'NONE':
+                        case 'RETRY':
+                            wf.status = 'RUNNING';
+                            break;
```
What I noticed is that the event details dialog will say 'RUNNING' even if the workflow has not been resumed yet by the WorkflowService (you can see its state in the events list in the background). Also, if the dialog remains open and the workflow fails, it will remain saying 'RUNNING". So I think we either make sure the state shown in the event details dialog is accurate or we need to close the dialog.

krinnewitz replied:
Ok, I think it is now acceptable to close the dialog since we don't have the pause button anymore and updating it correctly would mean a higher effort. I changed it back to closing the modal instead trying to update the content.

Commit:
Fix review findings

- fix ReST Annotations in AbstractEventEndpoint
- close modal when user clicks resume/abort/stop

rute-santos reviewed (approved):
Thank you for making the changes. Everything looks fine now!

krinnewitz commented:
Thank you for reviewing.
